apoc.coll.min() and max() do not work on temporal types

## Expected Behavior (Mandatory)
APOC max() and min() functions should work on temporal types

## Actual Behavior (Mandatory)
They don't. We get an error.

## How to Reproduce the Problem

### Simple Dataset (where it's possibile)
No dataset needed

### Steps (Mandatory)

`return apoc.coll.max([date(), date() - duration('p5M')])`
results in the following errro:

`Failed to invoke function `apoc.coll.max`: Caused by: java.lang.ClassCastException: java.time.LocalDate cannot be cast to java.lang.Number`
  

### Versions

  - OS: Mac OS Mojave
  - Neo4j: 3.5.16
  - Neo4j-Apoc: 3.5.0.9 (and presumably everything below, likely includes 4.0.0.x)

--- a/src/main/java/apoc/coll/Coll.java
+++ b/src/main/java/apoc/coll/Coll.java
@@ -6,6 +6,7 @@ import org.neo4j.graphdb.GraphDatabaseService;
 import org.neo4j.graphdb.Node;
 import org.neo4j.graphdb.Path;
 import org.neo4j.graphdb.Relationship;
+import org.neo4j.graphdb.Result;
 import org.neo4j.helpers.collection.Pair;
 import org.neo4j.kernel.impl.util.statistics.IntCounter;
 import org.neo4j.procedure.*;
@@ -89,18 +90,20 @@ public class Coll {
     @Description("apoc.coll.min([0.5,1,2.3])")
     public Object min(@Name("values") List<Object> list) {
 		if (list == null || list.isEmpty()) return null;
-        return Collections.min((List)list, Coll::compareAsDoubles);
+        if (list.size() == 1) return list.get(0);
+        try (Result result = db.execute("cypher expressionEngine=compiled return reduce(res=null, x in $list | CASE WHEN res IS NULL OR x<res THEN x ELSE res END) as value", Collections.singletonMap("list", list))) {
+            return result.next().get("value");
+        }
     }
 
     @UserFunction
     @Description("apoc.coll.max([0.5,1,2.3])")
     public Object max(@Name("values") List<Object> list) {
-		if (list == null || list.isEmpty()) return null;
-        return Collections.max((List)list, Coll::compareAsDoubles);
-    }
-
-    private static int compareAsDoubles(Object a, Object b) {
-        return Double.compare(((Number)a).doubleValue(), ((Number)b).doubleValue());
+        if (list == null || list.isEmpty()) return null;
+        if (list.size() == 1) return list.get(0);
+        try (Result result = db.execute("cypher expressionEngine=compiled return reduce(res=null, x in $list | CASE WHEN res IS NULL OR res<x THEN x ELSE res END) as value", Collections.singletonMap("list", list))) {
+            return result.next().get("value");
+        }
     }
 
     @Procedure
@@ -397,6 +400,7 @@ public class Coll {
     @UserFunction
     @Description("apoc.coll.indexOf(coll, value) | position of value in the list")
     public long indexOf(@Name("coll") List<Object> coll, @Name("value") Object value) {
+        // return reduce(res=[0,-1], x in $list | CASE WHEN x=$value AND res[1]=-1 THEN [res[0], res[0]+1] ELSE [res[0]+1, res[1]] END)[1] as value
         if (coll == null || coll.isEmpty()) return -1;
         return  new ArrayList<>(coll).indexOf(value);
     }

--- a/src/test/java/apoc/coll/CollTest.java
+++ b/src/test/java/apoc/coll/CollTest.java
@@ -105,6 +105,16 @@ public class CollTest {
     }
 
     @Test
+    public void testMaxDate() throws Exception {
+        testCall(db, "RETURN apoc.coll.max([date('2020-04-01'), date('2020-03-01')]) as value",
+                (row) -> assertEquals("2020-04-01", row.get("value").toString()));
+        testCall(db, "RETURN apoc.coll.max([datetime('2020-03-30T12:17:43.175Z'), datetime('2020-03-30T12:17:39.982Z')]) as value",
+                (row) -> assertEquals("2020-03-30T12:17:43.175Z", row.get("value").toString()));
+        testCall(db, "RETURN apoc.coll.max([null, datetime('2020-03-30T11:17:39.982Z'), datetime('2020-03-30T12:17:39.982Z'), datetime('2020-03-30T11:17:39.982Z')]) as value",
+                (row) -> assertEquals("2020-03-30T12:17:39.982Z", row.get("value").toString()));
+    }
+
+    @Test
     public void testPartition() throws Exception {
         testResult(db, "CALL apoc.coll.partition([1,2,3,4,5],2)",
                 (result) -> {

